### DISCUSSION TITLE
Git Worktrees support

### DISCUSSION THREAD

**Discussion by berkus** [zed-industries/zed#26084] [Feature Requests]
### Describe the feature

Similar to vscode gitlens extension, support checking out multiple branches into separate work folders and showing them all as multiple independent project trees within one window. (this uses `git worktree` commands underneath)

Will add some screeenshots in a bit. 

See: https://git-scm.com/docs/git-worktree

**Comment by notpeter** [comment]
Can you provide a simple example for how to build a mini project demonstrating this setup and how you would like to use it?

**Comment by berkus** [comment]
@notpeter So, in my repo github.com/metta-systems/vesper I have many branches for parallel workstreams. They are all parts of the same project but they are wildly different in the state of completion and even starting point (some are rebased, some are still off of some older trees).

## One approach I tried was:

1. create a .worktrees folder inside the "main" checkout that is open in Zed
2. inside the worktrees folder use "git worktree add" to create various subfolders, each corresponding to a particular branch/workstream/feature
3. open and edit files inside these subtrees from the single zed instance having the top directory with "main" checkout.

This is extremely unwieldy because of a few issues: 
1. the subtrees folders get randomly closed/collapsed based on some update logic somewhere, so i could be editing a file in a subtree and suddenly the folder I was in disappears, and i need to go to .worktrees and expand each level manually again
2. zed and rust-analyzer completely go bonkers because there are now some unrelated files inside the main checkout with completely separate Cargo.tomls and workspaces - editing code in this setup is a nightmare.
3. zed does not know that files inside .worktrees folder are actually different checkouts of the same repository, so it doesn't provide any assitance with those - all the worktree/git maintenance commands are manual via cli, the main zed windows shows "main" branch as checked out, even though i'm working in something like "feat/mmu-support".

## Another approach

I've tried opening multiple zed windows, each in one of the workspace subpaths of .worktrees folder. That works a bit better, but all maintenance steps are still manual, and there are many zed windows around even though I usually work on one worktree at a time.

## What am I expecting?

Native support for "git worktrees" - these are parts of git config, discoverable via git queries. Basically, I want to either replace branch selector or add one more selector next to it that would switch the active zed workspace to one of the defined git worktree locations (these can really be located ANYWHERE - as you specify the disk path for the subtree in the git cli). Also, I would like to see a native Zed support for creating a worktree - i.e. have a default path pattern where a new worktree would be created for branch X.

**Reply by berkus** [reply]
i.e. the listing of the current setup:

```
/somewhere/vesper.git <-- this is open in Zed and corresponds to "main" branch
+--.worktrees
      +--feat
            +--feat1.             <-- this is a git worktree registered for feat/feat1 branch
            +--feat2             <-- this is a git worktree registered for feat/feat1 branch
      +--fix
           +--bugfix1.        <-- this is a git worktree registered for fix/bugfix1 branch
```

The specific naming of the folders is not too important as far as:
1. they are not affecting each other too much (that's why .worktrees is a hidden dir - i don't want it to pollute the main workspace too much)
2. they are relatively independent and can be created and removed without affecting anything else (here it's not too good - ideally the worktree for "main" branch should also be inside .worktrees but zed doesn't support this configuration and it would be too awkward currently)

Ideally, the worktree support would work like one of the vscode worktree extensions does - it creates a BARE git repo checkout to track worktrees and then creates per-branch worktrees beside it, e.g.

```
vesper.git
+--.bare      <-- bare git repo, solely for shared repo storage
+--.worktrees
      +--main      <-- zed would pick that as just one of the worktrees and let switch to/from it as usual
      +--feat
           +--feat1 ....
```

**Comment by JacobCallahan** [comment]
I'm a big fan of using worktrees for larger features, or those that succumb to scope creep. Not sure if it will help with implementation ideas, but below is the bashrc function I made to work with worktrees.

All worktrees are stored in a directory next to the repo root directory.
That directory is named "repo_dir.worktrees".

```bash
# --- Interactive worktree manager ---
worktree() {
  # Handle help flag
  if [[ "$1" == "--help" ]] || [[ "$1" == "-h" ]]; then
    cat << 'EOF'
worktree - Interactive Git worktree manager

USAGE:
  worktree [name]           Create or switch to a worktree
  worktree                  List and select from existing worktrees
  worktree --help           Show this help message

DESCRIPTION:
  Manages Git worktrees in a dedicated directory structure outside the main
  repository. Worktrees are stored in <repo>.worktrees/ alongside the main repo.

EXAMPLES:
  # List all worktrees and select one interactively
  worktree

  # Create or switch to a worktree named "feature-branch"
  worktree feature-branch

  # When creating a new worktree, you'll be prompted to select which branch
  # to base it on using fzf

DIRECTORY STRUCTURE:
  /home/user/Programming/myrepo/              (main repository)
  /home/user/Programming/myrepo.worktrees/
    ├── feature-branch/                       (worktree)
    ├── bugfix/                               (worktree)
    └── experimental/                         (worktree)

NOTES:
  - When listing worktrees (no arguments), select "Exit" to return to main repo
  - Worktrees are automatically created in <repo>.worktrees/ directory
  - You can navigate between worktrees and the main repo seamlessly
EOF
    return 0
  fi

  # Determine the common .git directory path
  local common_git_dir
  common_git_dir=$(git rev-parse --git-common-dir 2>/dev/null)
  if [[ -z "$common_git_dir" ]]; then
    echo "System Alert: Operation must be performed within a Git repository or its worktree." >&2
    return 1
  fi

  # Determine if we're in a worktree or main repo
  local git_dir
  git_dir=$(git rev-parse --git-dir 2>/dev/null)

  # Determine the main repository root path - convert to absolute path
  local main_repo_root_path
  if [[ "$git_dir" == ".git" ]] || [[ "$git_dir" == */.git ]]; then
    # We're in the main repo - get absolute path
    main_repo_root_path=$(cd "$(git rev-parse --show-toplevel)" && pwd)
  else
    # We're in a worktree - common_git_dir points to main repo's .git
    main_repo_root_path=$(dirname "$common_git_dir")
  fi

  # Derive repository name and parent directory from the main repo root
  local repo_name
  repo_name=$(basename "$main_repo_root_path")
  local parent_dir
  parent_dir=$(dirname "$main_repo_root_path")
  local worktree_base_dir="${parent_dir}/${repo_name}.worktrees"

  # Simplified main branch detection (relative to the main repo context)
  local main_branch
  if git --git-dir="$common_git_dir" --work-tree="$main_repo_root_path" show-ref --verify --quiet refs/heads/main; then
      main_branch="main"
  else
      main_branch="master" # Assumes master if main doesn't exist
  fi
  local target_name="$1"

  if [[ -n "$target_name" ]]; then
    # Handle Provided Worktree Name
    local target_path="${worktree_base_dir}/${target_name}"
    # Check if the absolute path exists in the worktree list
    if git worktree list --porcelain | grep "^worktree " | cut -d ' ' -f 2- | grep -q "^${target_path}$"; then
      echo "Navigating to existing worktree: $target_path"
      cd "$target_path" || return 1
    else
      # Create new worktree
      echo "Worktree '$target_name' not found at expected location. Creating..."
      local branch_choice
      # List branches from the main repository context
      branch_choice=$(git --git-dir="$common_git_dir" --work-tree="$main_repo_root_path" branch --format='%(refname:short)' | fzf --height 15% --reverse --prompt="Select source branch for '$target_name': ")
      if [[ -n "$branch_choice" ]]; then
        # Ensure the base directory exists
        mkdir -p "$worktree_base_dir" || { echo "System Alert: Failed to create base directory $worktree_base_dir" >&2; return 1; }
        echo "Executing: git worktree add \"$target_path\" \"$branch_choice\""
        # Run git worktree add from the main repo context for consistency
        if git --git-dir="$common_git_dir" --work-tree="$main_repo_root_path" worktree add "$target_path" "$branch_choice"; then
           echo "Worktree created. Navigating to $target_path"
           cd "$target_path" || return 1
        else
           echo "System Alert: 'git worktree add' command failed." >&2
           # Simple cleanup attempt
           if [[ -d "$target_path" ]] && ! ls -A "$target_path" | grep -qv '.git'; then
               rmdir "$target_path" 2>/dev/null
           fi
           return 1
        fi
      else
        echo "Worktree creation cancelled."
        return 1
      fi
    fi
  else
    # Handle No Worktree Name Provided
    # List worktree *names* located in the designated external directory
    git worktree list --porcelain | grep "^worktree " | cut -d ' ' -f 2- | while read -r path; do
      # Skip the main repository and only show worktrees in the .worktrees directory
      if [[ "$path" == "$worktree_base_dir"/* ]]; then
        basename "$path"
      fi
    done

    local worktree_names
    worktree_names=$(git worktree list --porcelain | grep "^worktree " | cut -d ' ' -f 2- | while read -r path; do
      if [[ "$path" == "$worktree_base_dir"/* ]]; then
        basename "$path"
      fi
    done)

    local selection
    local exit_option="Exit (Switch to '$main_branch' in main directory)"
    local options="$exit_option\n$worktree_names"

    selection=$(echo -e "$options" | fzf --height 25% --reverse --prompt="Select worktree or Exit: ")

    if [[ "$selection" == "$exit_option" ]]; then
      # Use the correctly identified main repository root path
      echo "Navigating to main repository directory '$main_repo_root_path' and checking out '$main_branch'."
      cd "$main_repo_root_path" && git switch -f "$main_branch"
    elif [[ -n "$selection" ]]; then
      # Selection is a worktree name, reconstruct the path
      local selected_path="${worktree_base_dir}/${selection}"
      if [[ -d "$selected_path" ]]; then
         echo "Navigating to selected worktree: $selected_path"
         cd "$selected_path" || return 1
      else
         echo "System Alert: Selected worktree path '$selected_path' not found." >&2
         return 1
      fi
    else
       echo "No selection made."
    fi
  fi
  return 0
}
```

**Comment by medoror-mo** [comment]
Any updates on this? Imagine the desired functionality resembles this plugin that VS Code has for gitworktrees https://marketplace.visualstudio.com/items?itemName=GitWorktrees.git-worktrees

**Comment by fastjames** [comment]
Was this addressed in #4670 ?

**Reply by heksesang** [reply]
That issue only addressed the problem that you couldn't open a worktree directory in Zed with any sort of git support. This discussion is about making worktree-based workflows work more smoothly than they do today.

**Comment by heksesang** [comment]
Another issue steming from no worktree support in Zed is that if you do open a repository using worktrees in Zed and try to switch to a branch that you already have a worktree for, you get a very non-descriptive error that the checkout command failed. Not very good UX at all.

**Comment by aexvir** [comment]
with agentic development being more and more popular I'm finding worktrees to be essential for keeping parallel workstreams, so far I haven't seen any editor that embraces worktrees and offers a higher level control over these workstreams

this, in my mind, could align quite well with the vision that Nathan shared in the discussion with Steve, where perhaps collab channels could be used to define workstreams for agents to iterate on top of; if each channel would be backed up by its own worktree then features can be actually be built in parallel with little friction, be it with llms involved or not

for me an mvp of this would be some flag in the settings to let me manage branches as worktrees, so users that want will opt-in would get each branch in its own worktree; no changes on the ui, when switching branch it would switch worktree, creating and deleting branches would delete worktrees and so on

**Comment by jarlbrak** [comment]
> with agentic development being more and more popular I'm finding worktrees to be essential for keeping parallel workstreams, so far I haven't seen any editor that embraces worktrees and offers a higher level control over these workstreams
> 
> 
> 
> this, in my mind, could align quite well with the vision that Nathan shared in the discussion with Steve, where perhaps collab channels could be used to define workstreams for agents to iterate on top of; if each channel would be backed up by its own worktree then features can be actually be built in parallel with little friction, be it with llms involved or not
> 
> 
> 
> for me an mvp of this would be some flag in the settings to let me manage branches as worktrees, so users that want will opt-in would get each branch in its own worktree; no changes on the ui, when switching branch it would switch worktree, creating and deleting branches would delete worktrees and so on

Check out Crystal. I use it when there is enough ambiguity in my prompt to warrant exploring different solutions from parallel workflows.

**Comment by AlvaroParker** [comment]
Hi! I've opened a ~~draft~~ PR https://github.com/zed-industries/zed/pull/38719 although as a separate component to branches (similiar to how vscode does it for for `> worktree ...` commands).

**Comment by greg-truewind** [comment]
seems like with agentic development this is more needed as I see our engineers try to parallelize their workflows with multiple agents running. I think seamless switching between workspaces to help agent to progress on a task would be so powerful. conductor attempts to do this but it abstracts code editor entirely, I still want to be in the weeds of changing small code pieces as an agents run along in multiple worktrees

**Comment by plasticine** [comment]
Whatever gets built here it would be awesome if it could be generic or extensible enough to accomodate other worktree-ish workflows — [Workspaces](https://jj-vcs.github.io/jj/latest/working-copy/#workspaces) in [Jujutsu](https://jj-vcs.github.io/jj/latest/) for example 🙏

**Comment by berlund** [comment]
Started to use worktrees in a similar way as described here: https://gabri.me/blog/git-worktrees-done-right

However, opening a worktree folder or even the folder containing all worktrees lets the git integration assume there is no branch selected. Having support for worktrees for at least making 'common' git operations would be really nice!

**Comment by zmanian** [comment]
I think the next step here is less about adding more standalone worktree actions and more about making worktrees part of Zed’s core workspace/git/agent model.

The worktree picker and create/open flow were a meaningful improvement, but there still seems to be a mismatch between:

- Git state that is fundamentally worktree-scoped
- UI/actions that are still largely branch- or repo-scoped
- agent/editor workflows that can involve multiple sibling worktrees from the same underlying repo

From a product/architecture perspective, I think the missing abstraction is something like an explicit “active worktree context” that is durable and shared across subsystems, rather than being a temporary UI selection.

Concretely, I think that would unlock better behavior in a few places:

- Branch switching:
  - if a branch is already checked out in another worktree, branch-switch flows should be able to resolve to “open/switch to that worktree” instead of only attempting checkout in the current one
  - if not, the UX could offer checkout-in-place vs create-worktree, with the latter being a first-class path rather than a separate command
- Workspace state:
  - the selected worktree context should remain stable until the user intentionally changes it
  - other workspace actions should be able to consult that context instead of inferring from whichever editor/repo most recently became active
- Git UI:
  - repo-level UI could expose the set of worktrees as first-class state: branch, path, dirty state, ahead/behind, and open/remove/create actions
  - this would make the current worktree support discoverable and also reduce ambiguity when multiple sibling worktrees exist
- Agent context:
  - the agent should have explicit access to the current worktree context
  - search, edits, diagnostics, terminals, and git actions should all be scoped to that worktree unless the user intentionally broadens scope
  - “work on this branch/task in a new worktree” seems like a natural agent primitive, especially for parallel or issue-based workflows

The agent point feels especially important. Worktrees are one of the cleanest isolation boundaries for parallel development, and if Zed is leaning into agentic workflows, then worktree-awareness seems like the right model boundary rather than repo-only awareness.

So my suggestion would be to frame future work here around a shared worktree-context model, with branch switching, git UI, and agent scope all consuming that same state. That feels more likely to produce a coherent UX than continuing to add isolated worktree commands.

**Reply by tredondo** [reply]
\+ Ability to rename worktrees and updating agents threads to match the rename. `git worktree move` will leave agent threads orphan.

**Comment by DongDongBear** [comment]
In any case, please prioritize implementing this soon. I really need to be able to see the changes made by Claude Code / Codex in worktrees directly from the Git panel.

**Comment by lessevv** [comment]
I really want to switch over to Zed as I'm loving it, but the current worktree flow feels like it breaks certain flows (for example, I use Get Shit Done with worktree mode, and it seems like switching to the agents worktree breaks the subagent flow). Would be amazing if from the orchestrator session, it also follows the changes from the worktrees created.

**Comment by 4nur4g** [comment]
This is the only missing feature that's stopping me to completely switch to Zed. I'm stuck with VS Code for now. I'm eagerly waiting for it to be implemented.

**Comment by zengzpi** [comment]
I love Zed so much that I’ve completely abandoned VS Code, Cursor, and every other IDE for it! To bring over some of my favorite workflows, I tried Gemini 3.5 Flash to build a working proof-of-concept for a simultaneous, stacked Git Worktrees view in the sidebar.

Instead of making you juggle global context switches via title-bar dropdowns, this implementation dynamically stacks active checkouts vertically. Each worktree acts as an independent, collapsible container featuring its own dedicated Staged, Unstaged, and Untracked sections, alongside a localized commit input box. To top it off, I also replaced the default "History" tab with an integrated, visual **Git Graph** panel for seamless multi-branch tracking.

<img width="528" height="729" alt="image" src="https://github.com/user-attachments/assets/794edd3e-410a-4136-8c44-00b70c3a198a" />
<img width="558" height="366" alt="image" src="https://github.com/user-attachments/assets/eda30493-0684-4b2d-9ab2-7f56544541e4" />

I'd love to get everyone's thoughts on this! Maintainers, if you're open to landing this architectural update, I’ll gladly clean up my branch and submit a formal PR.

**Comment by l2ysho** [comment]
Hey, I got 2 small things I would like to point attention. (Both land on the point made upthread).

**1. Checking out a branch that lives in another worktree is a dead end.**

`git: checkout branch` only ever attempts a checkout in the current worktree, so git refuses and the raw stderr lands in a modal whose only button is OK:

> Failed to change branch
> Git command failed:
> fatal: 'claude/apify-cli-1106-hotfix-c69e50' is already used by worktree at '/Users/me/Work/apify-cli/.claude/worktrees/pnpm-store-sandbox-issue-d9175c'

**2. You can't find a worktree by branch name.**

this is especially annoying when woktrees are created by agent with random names.

Both seems simple to do, I am happy to open separate issues or send a PR for either if that's wanted.

**Comment by mkeeler** [comment]
In the worktree selection, Zed seems to assume that the last path element should be the name of the worktree but that is not always the case. I tend to have worktrees that are more nested like:

* `mkeeler/<some feature>` - worktree I am actively working on
* `<other user>/<some feature>` - code I am reviewing
* `release/X.Y.Z` - release branches

It would be great if Zed could infer the worktree name starting at the configured worktree directory path. So if all my worktrees are created in `/some/path` with a particular worktree at `/some/path/mkeeler/foo` then the worktree name would be `mkeeler/foo` and not just `foo`.

**Reply by matthewgrossman** [reply]
https://github.com/zed-industries/zed/discussions/60068 is related

**Comment by meza** [comment]
Also absolute paths would be great! I keep all my worktrees in one specific folder on a specific drive. Not being able to do that really hurts
